### ISSUE TITLE
An Empirical Evaluation of Generic Convolutional and Recurrent Networks for Sequence Modeling

### ISSUE THREAD

**Issue by jinglescode** [jinglescode/ml-papers#14]
# Paper
Link: https://arxiv.org/pdf/1803.01271.pdf
Year: 2018

# Summary
Temporal Convolutional Networks, or simply TCN is a variation over Convolutional Neural Networks for sequence modelling tasks. Rather, it’s quite a descriptive term for a family of architectures.

Motivation:
- TCNs exhibit longer memory than recurrent architectures with the same capacity.
- Constantly performs better than LSTM/GRU architectures on a vast range of tasks (Seq. MNIST, Adding Problem, Copy Memory, Word-level PTB...).
- Parallelism, flexible receptive field size, stable gradients, low memory requirements for training, variable length inputs.

Some distinguishable characteristics for TCNs are:
1. The convolutions in the architecture are causal(there is no information leakage) from future to past.
2. The architecture can take a sequence of any length and map it to an output sequence of the same length, just as with an Recurrent Neural Networks.

Key points:
- Sequence Modeling: This formalism encompasses many settings such as autoregressive prediction (where we try to predict some signal given its past) by setting the target output to be simply the input shifted by one time step.
- Causal Convolutions: As mentioned above, the TCN is based upon two principles: the fact that there can be no leakage from the future into the past and the fact that the network produces an output of the same length as the input. To accomplish the first point, the TCN uses causal convolutions, convolutions where an output at time t is convolved only with elements from time t and earlier in the previous layer. To achieve the second point, the TCN uses a 1D fully-convolutional network (FCN) architecture, where each hidden layer is the same length as the input layer, and zero padding of length (kernel size − 1) is added to keep subsequent layers the same length as previous ones.
- TCNs do not use gating mechanisms and have much longer memory.

Advantages:
- Parallelism: Unlike in RNNs where the predictions for later timesteps must wait for their predecessors to complete, convolutions can be done in parallel since the same filter is used in each layer. Therefore, in both training and evaluation, a long input sequence can be processed as a whole in TCN, instead of sequentially as in RNN.
- Flexible receptive field size: A TCN can change its receptive field size in multiple ways. For instance, stacking more dilated (causal) convolutional layers, using larger dilation factors, or increasing the filter size are all viable options (with possibly different interpretations). TCNs thus afford better control of the model’s memory size, and are easy to adapt to different domains.
- Stable gradients: Unlike recurrent architectures, TCN has a backpropagation path different from the temporal direction of the sequence. TCN thus avoids the problem of exploding/vanishing gradients, which is a major issue for RNNs (and which led to the development of LSTM, GRU, HF-RNN, etc)
- Low memory requirement for training: Especially in the case of a long input sequence, LSTMs and GRUs can easily use up a lot of memory to store the partial results for their multiple cell gates.

Disadvantages:
- Data storage during evaluation: In evaluation/testing, RNNs only need to maintain a hidden state and take in a current input  𝑥𝑡 in order to generate a prediction. In other words, a “summary” of the entire history is provided by the fixed-length set of vectors ℎ𝑡,  and the actual observed sequence can be discarded. In contrast, TCNs need to take in the raw sequence up to the effective history length, thus possibly requiring more memory during evaluation.
- Potential parameter change for a transfer of domain: Different domains can have different requirements on the amount of history the model needs in order to predict. Therefore, when transferring a model from a domain where only little memory is needed (i.e., small k and d) to a domain where much longer memory is required (i.e., much larger k and d), TCN may perform poorly for not having a sufficiently large receptive field.

- https://github.com/locuslab/TCN
- https://medium.com/the-artificial-impostor/notes-understanding-tensorflow-part-3-7f6633fcc7c7
